ppoprf/src/ppoprf.rs

```diff
@@ -37,351 +37,392 @@ pub const DIGEST_LEN: usize = 64;
-        let p = p.0;
-        let point = p.decompress().unwrap();
-        if self.public_key.get(md).is_none() {
-            return Err(PPRFError::BadTag { md });
-        }
-        let mut tag = [0u8; 32];
-        self.pprf.eval(&[md], &mut tag)?;
-        let ts = Scalar::from_bytes_mod_order(tag);
-        let tagged_key = self.oprf_key + ts;
-        let exponent = tagged_key.invert();
-        let eval_point = exponent * point;
-        let mut proof = None;
-        if verifiable {
-            let public_value = self.public_key.get_combined_pk_value(md)?;
-            proof = Some(ProofDLEQ::new(
-                &tagged_key,
-                &public_value,
-                &eval_point,
-                &point,
-            ));
-        }
-        Ok(Evaluation {
-            output: eval_point.compress(),
-            proof,
-        })
+    Ok(Self {
+      oprf_key,
+      public_key: ServerPublicKey {
+        base_pk: Point::from(oprf_key * RISTRETTO_BASEPOINT_POINT),
+        md_pks,
+      },
+      pprf,
+    })
+  }
+
+  pub fn eval(
+    &self,
+    p: &Point,
+    md: u8,
+    verifiable: bool,
+  ) -> Result<Evaluation, PPRFError> {
+    let p = p.0;
+    let point = p.decompress().unwrap();
+    if self.public_key.get(md).is_none() {
+      return Err(PPRFError::BadTag { md });
     }
-
-    pub fn puncture(&mut self, md: u8) -> Result<(), PPRFError> {
-        self.pprf.puncture(&[md])
-    }
-
-    pub fn get_public_key(&self) -> ServerPublicKey {
-        self.public_key.clone()
+    let mut tag = [0u8; 32];
+    self.pprf.eval(&[md], &mut tag)?;
+    let ts = Scalar::from_bytes_mod_order(tag);
+    let tagged_key = self.oprf_key + ts;
+    let exponent = tagged_key.invert();
+    let eval_point = exponent * point;
+    let mut proof = None;
+    if verifiable {
+      let public_value = self.public_key.get_combined_pk_value(md)?;
+      proof = Some(ProofDLEQ::new(
+        &tagged_key,
+        &public_value.into(),
+        &eval_point,
+        &point,
+      ));
     }
+    Ok(Evaluation {
+      output: Point(eval_point.compress()),
+      proof,
+    })
+  }
+
+  pub fn puncture(&mut self, md: u8) -> Result<(), PPRFError> {
+    self.pprf.puncture(&[md])
+  }
+
+  pub fn get_public_key(&self) -> ServerPublicKey {
+    self.public_key.clone()
+  }
 }
 
 // The `Client` struct is essentially a collection of static functions
 // for computing client-side operations in the PPOPRF protocol.
 pub struct Client {}
 impl Client {
-    pub fn blind(input: &[u8]) -> (Point, Scalar) {
-        let mut hashed_input = [0u8; 64];
-        strobe_hash(input, "ppoprf_derive_client_input", &mut hashed_input);
-        let point = RistrettoPoint::from_uniform_bytes(&hashed_input);
-        let mut csprng = OsRng;
-        let r = Scalar::random(&mut csprng);
-        (Point((r * point).compress()), r)
+  pub fn blind(input: &[u8]) -> (Point, [u8; 32]) {
```

Comment:
I guess the primary motivation for `Point` is so that we can support more ciphersuites. In the [standardisation doc for OPRFs](https://github.com/cfrg/draft-irtf-cfrg-voprf), there are multiple ciphersuites that we could support. For instance, ideally we would support an elliptic curve that is a little bigger for cryptanalysis reasons (but I used ristretto because the rust lib is really nice and easy to use). Accordingly, if we wanted to make that change, the plan is that only exposing `Point` would mean that we would only have to change the ppoprf crate, and that sta-rs (and other apps) would go unimpacted. Personally, I think there is some worth in maintaining this flexibility, and similarly for the `Scalar` type which is also dependent on the ristretto curve.

Does this make sense? Or do you remain unconvinced?

Reply:
I'm unconvinced that `[u8; 32]` is the type we should be using here. I'd like to see some named type, whether it's the `Scalar` from curve25519_dalek we're using now, or something specific to the ppoprf crate like `Point`.

If you used a larger curve, would 32 bytes still be the correct size for the blinding token? Would a different ciphersuite do something besides scalar multiplication for the blind/unblind operation, so `Scalar` isn't a sufficiently general name? Wrapper types have no runtime cost in Rust, so it's good to use them whenever reasonable to get type-checking on function arguments. What other constraints are motivating using an array instead?

Reply:
Sorry yeah I forgot to respond to this in my last comment. In the latest change I have actually implemented `CurveScalar` as a wrapper type for this array, does this look better to you?

Reply:
Yes, that solution is fine. Thanks!